### PR TITLE
Implement vBAT expiry countdown

### PR DESCRIPTION
**Pull request by deeppandya**
Resolves https://github.com/brave/internal/issues/978
Resolves https://github.com/brave/brave-browser/issues/27883

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java [hunk 1]
```diff
@@ -1064,6 +1085,7 @@ public void OnNotificationDeleted(String id) {
 
     @Override
     public void OnGetExternalWallet(String externalWallet) {
+        Log.e("NTP", "OnGetExternalWallet : " + externalWallet);
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1088]: not needed

File: android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java [hunk 2]
```diff
@@ -1601,6 +1623,71 @@ public void onClick(View v) {
...
+
+        String vbatModalTitleString = "";
+        String vbatModalTextString = "";
+        double dueDateInMillis = mBraveRewardsNativeWorker.getVbatDeadline();
+        SimpleDateFormat sdf =
+                new SimpleDateFormat("MMMM dd, yyyy 'at' h:mm a", Locale.getDefault());
+        String dueDate = sdf.format(new Date((long) dueDateInMillis))
+                                 .replace("a.m.", "AM")
+                                 .replace("p.m.", "PM");
+        ;
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1676]: not needed

File: android/java/res/layout/rewards_panel_vbat_expire_notice.xml [hunk 3]
```diff
@@ -0,0 +1,145 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright 2023 The Brave Authors. All rights reserved.
+     Use of this source code is governed by a BSD-style license that can be
+     found in the LICENSE file. -->
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/rewards_panel_vbat_expire_notice.xml:4]: the header is incorrect

File: android/java/res/layout/rewards_panel_vbat_expire_notice.xml [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+			        <FrameLayout
+			        	android:id="@+id/btn_vbat_connect_account"
+			        	android:layout_width="match_parent"
+			        	android:layout_height="wrap_content"
+			        	android:layout_marginHorizontal="24dp"
+			        	android:layout_marginTop="16dp"
+			        	android:background="@drawable/blue_48_rounded_bg">
+
+				        <android.widget.Button
+		                    android:layout_width="wrap_content	"
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/rewards_panel_vbat_expire_notice.xml:91]: what happened with a format in that file?

File: android/java/res/layout/rewards_panel_vbat_expire_notice.xml [hunk 5]
```diff
@@ -0,0 +1,145 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c) 2020 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/res/layout/rewards_panel_vbat_expire_notice.xml:2]: why 2020? is that file new?

Comment (deeppandya) [reply android/java/res/layout/rewards_panel_vbat_expire_notice.xml:2]: oops. copy-paste error. Sorry my bad.

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java [hunk 6]
```diff
@@ -1602,6 +1623,70 @@ public void onClick(View v) {
...
+        }));
+
+        String vbatModalTitleString = "";
+        String vbatModalTextString = "";
+        double dueDateInMillis = mBraveRewardsNativeWorker.getVbatDeadline();
+        SimpleDateFormat sdf =
+                new SimpleDateFormat("MMMM dd, yyyy 'at' h:mm a", Locale.getDefault());
+        String dueDate = sdf.format(new Date((long) dueDateInMillis))
+                                 .replace("a.m.", "AM")
+                                 .replace("p.m.", "PM");
```
Comment (emerick) [inline android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1675]: AM/PM doesn't need to be localized?

**Review by emerick — APPROVED** [review]
LGTM, just had one localization-related question.

File: android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java (hunk 6 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1675]: Do we localized those in desktop ? i am not sure.

Comment (emerick) [reply android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1675]: I don't recall ever having to worry about it, to be honest. It just sort of stuck out to me with the calls to `replace`. Might be worth double-checking. Would there even be an "a.m." or "p.m." to replace in non-English locales?

File: android/java/res/layout/rewards_panel_vbat_expire_notice.xml [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+                    <TextView
+                        android:id="@+id/vbat_modal_text"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:layout_gravity="center"
+                        android:layout_marginHorizontal="24dp"
+                        android:layout_marginTop="24dp"
+                        android:fontFamily="@font/poppins_400"
+                        android:gravity="center"
```
Comment (sujitacharya2005) [inline android/java/res/layout/rewards_panel_vbat_expire_notice.xml:77]: hope anyone is sufficient  android:layout_gravity="center" and android:gravity="center"  OR android:textAlignment="viewStart"

File: android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java (hunk 6 above)
Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1675]: tbh we shouldn't do this. we should use the locale default format.

Comment (zenparsing) [reply android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1675]: Will the `'at'` be localized correctly, or will that just be inserted in the resulting string? If it's not localized, then we'll need to use a different way to format the date, I suppose.

Comment (deeppandya) [reply android/java/org/chromium/chrome/browser/rewards/BraveRewardsPanel.java:1675]: @zenparsing in that case, we should use default android date formatting.

**Pull request merged**